kimchi/src/tests/not.rs

```diff
@@ -0,0 +1,451 @@
+use o1_utils::{BigUintHelpers, BitwiseOps, FieldHelpers, RandomField};
+use rand::{rngs::StdRng, SeedableRng};
+
+type PallasField = <Pallas as AffineCurve>::BaseField;
+type VestaField = <Vesta as AffineCurve>::BaseField;
+type SpongeParams = PlonkSpongeConstantsKimchi;
+type VestaBaseSponge = DefaultFqSponge<VestaParameters, SpongeParams>;
+type VestaScalarSponge = DefaultFrSponge<Fp, SpongeParams>;
+type PallasBaseSponge = DefaultFqSponge<PallasParameters, SpongeParams>;
+type PallasScalarSponge = DefaultFrSponge<Fq, SpongeParams>;
+
+const NOT: bool = false;
+
+const RNG_SEED: [u8; 32] = [
+    211, 31, 143, 75, 29, 255, 0, 126, 237, 193, 86, 160, 1, 90, 131, 221, 186, 168, 4, 95, 50, 48,
+    89, 29, 13, 250, 215, 172, 130, 24, 164, 162,
+];
+
+// Constraint system for Not gadget using Xor16
+fn create_test_constraint_system_not_xor<G: KimchiCurve, EFqSponge, EFrSponge>(
+    bits: usize,
+) -> ConstraintSystem<G::ScalarField>
+where
+    G::BaseField: PrimeField,
+{
+    let (mut next_row, mut gates) = {
+        let mut gates = vec![CircuitGate::<G::ScalarField>::create_generic_gadget(
+            Wire::for_row(0),
+            GenericGateSpec::Pub,
+            None,
+        )];
+        let next_row = CircuitGate::<G::ScalarField>::extend_not_xor_gadget(&mut gates, 0, 1, bits);
+        (next_row, gates)
+    };
+
+    // Temporary workaround for lookup-table/domain-size issue
+    for _ in 0..(1 << 13) {
+        gates.push(CircuitGate::zero(Wire::for_row(next_row)));
+        next_row += 1;
+    }
+
+    ConstraintSystem::create(gates).public(1).build().unwrap()
+}
+
+// Constraint system for Not gadget using generic gates
+fn create_test_constraint_system_not_gnrc<G: KimchiCurve, EFqSponge, EFrSponge>(
+    nots: usize,
+) -> ConstraintSystem<G::ScalarField>
+where
+    G::BaseField: PrimeField,
+{
+    let mut gates = vec![CircuitGate::<G::ScalarField>::create_generic_gadget(
+        Wire::for_row(0),
+        GenericGateSpec::Pub,
+        None,
+    )];
+    let mut next_row =
+        CircuitGate::<G::ScalarField>::extend_not_gnrc_gadget(&mut gates, nots, 0, 1);
+
+    // Temporary workaround for lookup-table/domain-size issue
+    for _ in 0..(1 << 13) {
+        gates.push(CircuitGate::zero(Wire::for_row(next_row)));
+        next_row += 1;
+    }
+
+    ConstraintSystem::create(gates).public(1).build().unwrap()
+}
+
+// Creates the witness and circuit for NOT gadget using XOR
+fn setup_not_xor<G: KimchiCurve, EFqSponge, EFrSponge>(
+    inp: Option<G::ScalarField>,
+    bits: Option<usize>,
+) -> (
+    [Vec<G::ScalarField>; COLUMNS],
+    ConstraintSystem<G::ScalarField>,
+)
+where
+    G::BaseField: PrimeField,
+    EFqSponge: Clone + FqSponge<G::BaseField, G, G::ScalarField>,
+    EFrSponge: FrSponge<G::ScalarField>,
+{
+    let rng = &mut StdRng::from_seed(RNG_SEED);
+
+    let inp = rng.gen(inp, bits);
+
+    // If user specified a concrete number of bits, use that (if they are sufficient to hold the input)
```

Comment:
Ah, OK - now I see how the optional bits is used.  Maybe I missed it earlier, but if not, please comment the signature of those functions accordingly, so it's clear to the user.